In [100]:
import pandas as pd

In [101]:
url = "https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv"
df = pd.read_csv(url)
df.head()

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True


นี่คือชุดข้อมูล ไททานิค

ผมจะอธิบายเกี่ยวกับชุดข้อมูลก่อน

1.survived( Teaget)
0 = ไม่รอด
1 = รอด

2.pclass(อันดับคลาส)
1 = First class
2 = Second class
3 = Third class

3.sex
male
female

4.age

5.sibsp จำนวนพี่น้อง + คู่สมรสที่เดินทางมาด้วย
0 = เดินทางคนเดียว
1 = มากับพี่น้องหรือคู่สมรส 1 คน
2 = มากับ 2 คน
3 = มากับ 3 คน
4 = มากับ 4 คน
5 = มากับ 5 คน
8 = มากับ 8 คน

6.parch = จำนวนพ่อแม่ + ลูกที่มาด้วย
0 = ไม่มีพ่อแม่หรือลูกมาด้วย
1 = มากับลูกหรือพ่อแม่ 1 คน
2 = มากับ 2 คน
3 = มากับ 3 คน
...
6 = มากับ 6 คน

7.fare ราคาตั๋วที่จ่าย

8.embarked ท่าเรือที่ขึ้นเรือ ขึ้นเรือจากเมืองไหน
C = Cherbourg
Q = Queenstown
S = Southampton

9.class เหมือน pclass แต่เป็นข้อความแทนตัวเลข

10.who บอกประเภทคน

11.adult_male บอกว่าเป็น ผู้ชายที่เป็นผู้ใหญ่

12.deck ชั้นของดาดฟ้าเรือ

13.embark_town ชื่อเมืองเต็มของ embarked

14.alive เหมือน survived แต่เป็นข้อความ

15.alone บอกว่าเดินทาง








In [102]:
for column in df.columns:
    print(f"Unique values in column '{column}':")
    print(df[column].unique())
    print("\n" + "-"*30 + "\n")

Unique values in column 'survived':
[0 1]

------------------------------

Unique values in column 'pclass':
[3 1 2]

------------------------------

Unique values in column 'sex':
['male' 'female']

------------------------------

Unique values in column 'age':
[22.   38.   26.   35.     nan 54.    2.   27.   14.    4.   58.   20.
 39.   55.   31.   34.   15.   28.    8.   19.   40.   66.   42.   21.
 18.    3.    7.   49.   29.   65.   28.5   5.   11.   45.   17.   32.
 16.   25.    0.83 30.   33.   23.   24.   46.   59.   71.   37.   47.
 14.5  70.5  32.5  12.    9.   36.5  51.   55.5  40.5  44.    1.   61.
 56.   50.   36.   45.5  20.5  62.   41.   52.   63.   23.5   0.92 43.
 60.   10.   64.   13.   48.    0.75 53.   57.   80.   70.   24.5   6.
  0.67 30.5   0.42 34.5  74.  ]

------------------------------

Unique values in column 'sibsp':
[1 0 3 4 2 5 8]

------------------------------

Unique values in column 'parch':
[0 1 2 5 3 4 6]

------------------------------

Unique valu

# **ดูภาพรวมข้อมูล**

In [103]:
df.shape

(891, 15)

In [104]:
df.columns

Index(['survived', 'pclass', 'sex', 'age', 'sibsp', 'parch', 'fare',
       'embarked', 'class', 'who', 'adult_male', 'deck', 'embark_town',
       'alive', 'alone'],
      dtype='object')

In [105]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     891 non-null    int64  
 1   pclass       891 non-null    int64  
 2   sex          891 non-null    object 
 3   age          714 non-null    float64
 4   sibsp        891 non-null    int64  
 5   parch        891 non-null    int64  
 6   fare         891 non-null    float64
 7   embarked     889 non-null    object 
 8   class        891 non-null    object 
 9   who          891 non-null    object 
 10  adult_male   891 non-null    bool   
 11  deck         203 non-null    object 
 12  embark_town  889 non-null    object 
 13  alive        891 non-null    object 
 14  alone        891 non-null    bool   
dtypes: bool(2), float64(2), int64(4), object(7)
memory usage: 92.4+ KB


In [106]:
df.describe()

,survived,pclass,age,sibsp,parch,fare
count,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


In [107]:
df.groupby("age")["survived"].mean()

,survived
age,
0.42,1.0
0.67,1.0
0.75,1.0
0.83,1.0
0.92,1.0
...,...
70.00,0.0
70.50,0.0
71.00,0.0


# **วิเคราะห์ target คือsurvived**
เพราะว่าอยากรู้ว่า
- รอดกี่คน
- ไม่รอดกี่คน
- สัดส่วนประมาณเท่าไร

In [108]:
df["survived"].value_counts()

,count
survived,
0,549
1,342


In [109]:
df["survived"].value_counts(normalize=True)

,proportion
survived,
0,0.616162
1,0.383838


# **วิเคราะห์ feature สำคัญ**
โดยที่ผมจะแบบออกเป็นหมวด เพื่อให้เข้าใจง่าย ว่าตัวแปรไหนที่สำคัญกับ Y บ้าง แบ่งออกได้แก่ Numeric Features, Categorical Features, Boolean Features

In [110]:
numeric_features = df.select_dtypes(include=["int64","float64"]).columns
print(numeric_features)

Index(['survived', 'pclass', 'age', 'sibsp', 'parch', 'fare'], dtype='object')


In [111]:
for col in numeric_features:
    print(f"\n=== {col} ===")
    print(df.groupby(col)["survived"].mean())


=== survived ===
survived
0    0.0
1    1.0
Name: survived, dtype: float64

=== pclass ===
pclass
1    0.629630
2    0.472826
3    0.242363
Name: survived, dtype: float64

=== age ===
age
0.42     1.0
0.67     1.0
0.75     1.0
0.83     1.0
0.92     1.0
        ... 
70.00    0.0
70.50    0.0
71.00    0.0
74.00    0.0
80.00    1.0
Name: survived, Length: 88, dtype: float64

=== sibsp ===
sibsp
0    0.345395
1    0.535885
2    0.464286
3    0.250000
4    0.166667
5    0.000000
8    0.000000
Name: survived, dtype: float64

=== parch ===
parch
0    0.343658
1    0.550847
2    0.500000
3    0.600000
4    0.000000
5    0.200000
6    0.000000
Name: survived, dtype: float64

=== fare ===
fare
0.0000      0.066667
4.0125      0.000000
5.0000      0.000000
6.2375      0.000000
6.4375      0.000000
              ...   
227.5250    0.750000
247.5208    0.500000
262.3750    1.000000
263.0000    0.500000
512.3292    1.000000
Name: survived, Length: 248, dtype: float64


In [112]:
df[numeric_features].corr()

,survived,pclass,age,sibsp,parch,fare
survived,1.000000,-0.338481,-0.077221,-0.035322,0.081629,0.257307
pclass,-0.338481,1.000000,-0.369226,0.083081,0.018443,-0.549500
age,-0.077221,-0.369226,1.000000,-0.308247,-0.189119,0.096067
sibsp,-0.035322,0.083081,-0.308247,1.000000,0.414838,0.159651
parch,0.081629,0.018443,-0.189119,0.414838,1.000000,0.216225
fare,0.257307,-0.549500,0.096067,0.159651,0.216225,1.000000


In [113]:
categorical_features = df.select_dtypes(include=["object","category"]).columns
print(categorical_features)

Index(['sex', 'embarked', 'class', 'who', 'deck', 'embark_town', 'alive'], dtype='object')


In [114]:
for col in categorical_features:
    print(f"\n=== {col} ===")
    print(df[col].value_counts())
    print(df.groupby(col)["survived"].mean())


=== sex ===
sex
male      577
female    314
Name: count, dtype: int64
sex
female    0.742038
male      0.188908
Name: survived, dtype: float64

=== embarked ===
embarked
S    644
C    168
Q     77
Name: count, dtype: int64
embarked
C    0.553571
Q    0.389610
S    0.336957
Name: survived, dtype: float64

=== class ===
class
Third     491
First     216
Second    184
Name: count, dtype: int64
class
First     0.629630
Second    0.472826
Third     0.242363
Name: survived, dtype: float64

=== who ===
who
man      537
woman    271
child     83
Name: count, dtype: int64
who
child    0.590361
man      0.163873
woman    0.756458
Name: survived, dtype: float64

=== deck ===
deck
C    59
B    47
D    33
E    32
A    15
F    13
G     4
Name: count, dtype: int64
deck
A    0.466667
B    0.744681
C    0.593220
D    0.757576
E    0.750000
F    0.615385
G    0.500000
Name: survived, dtype: float64

=== embark_town ===
embark_town
Southampton    644
Cherbourg      168
Queenstown      77
Name: count, dt

In [115]:
bool_features = df.select_dtypes(include=["bool"]).columns
print(bool_features)

Index(['adult_male', 'alone'], dtype='object')


In [116]:
for col in bool_features:
    print(f"\n=== {col} ===")
    print(df.groupby(col)["survived"].mean())


=== adult_male ===
adult_male
False    0.717514
True     0.163873
Name: survived, dtype: float64

=== alone ===
alone
False    0.505650
True     0.303538
Name: survived, dtype: float64


# **Insight**

พบว่า
1.เพศมีผลต่อการรอดชีวิตอย่างมาก เพราะว่าในเหตุการณ์ Titanic มีนโยบายช่วยเหลือ “women and children first”
female survival ≈ 74%
male survival ≈ 19%

2.ผู้โดยสารชั้น 1 มีโอกาสรอดสูงกว่าชั้น 3 อย่างมาก เพราะว่าผู้โดยสารชั้นสูงอยู่ใกล้ดาดฟ้าและเรือชูชีพมากกว่า
First class survival สูง
Third class survival ต่ำ

3.ค่าโดยสารสัมพันธ์กับการรอด

fare สูง → survival สูง
โดยที่ ผู้โดยสารที่จ่ายค่าตั๋วสูงมีแนวโน้มรอดมากกว่า เพราะว่า ค่าโดยสารสะท้อนสถานะทางเศรษฐกิจ และมักอยู่ในชั้นโดยสารที่ดีกว่า

4.ขนาดครอบครัวมีผลต่อ survival
family_size เล็ก → survival สูงกว่า
เพราะ ผู้โดยสารที่มากับครอบครัวขนาดเล็กมีโอกาสรอดมากกว่าครอบครัวใหญ่เหตุผลคือครอบครัวใหญ่เคลื่อนที่และอพยพได้ยากกว่า

5.ผู้โดยสารที่เดินทางคนเดียวมี pattern ต่างจากผู้ที่มากับครอบครัว
is_alone = True มี survival ต่างจาก False
การเดินทางคนเดียวหรือมากับครอบครัวอาจส่งผลต่อโอกาสรอด เพราะคนที่มีครอบครัวอาจช่วยกันหรือรอกัน ทำให้การอพยพช้าลง

6.บาง feature มี missing values จำนวนมากเช่น age
deck ข้อมูลบางส่วนมี missing values สูง ซึ่งต้องจัดการก่อนสร้างโมเดล เพราะmissing data อาจทำให้โมเดลเรียนรู้ผิดพลาดได้





# **Data Cleaning**

**MISSING VALUES**

In [117]:
df.isnull().sum()

,0
survived,0
pclass,0
sex,0
age,177
sibsp,0
parch,0
fare,0
embarked,2
class,0
who,0


age มีข้อมูลหายไป 177 คน

deck ชั้นของดาดฟ้าเรือ เช่น A
B
C
D หายไป 688 แถว

embark_town ชื่อเมืองเต็มของ embarked เช่น Southampton
Cherbourg
Queenstown
หายไป 2แถว




**วิเคราะห์ทีละตัว**
แนวคิดคือ ยังสำคัญ → เติมค่า

age

In [118]:
df["age"] = df["age"].fillna(df["age"].median())  # missing ≈ 20% เติมค่ากลาง

deck missing ≈ 77%

In [119]:
df["deck_missing"] = df["deck"].isnull().astype(int)

 embark_town 2 แถว

In [120]:
df["embarked"].fillna(df["embarked"].mode()[0], inplace=True)

/tmp/ipykernel_1997/3995229981.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df["embarked"].fillna(df["embarked"].mode()[0], inplace=True)


# **Feature Engineering**

STEP 1: สร้าง feature ใหม่

1) family_size
เหตุผล: sibsp กับ parch แยกกันอาจเห็นภาพไม่ชัด รวมเป็น “ขนาดครอบครัวบนเรือ” จะตีความง่ายกว่า

In [121]:
df["family_size"] = df["sibsp"] + df["parch"] + 1

In [122]:
df["family_size"].value_counts()
df.groupby("family_size")["survived"].mean()

,survived
family_size,
1,0.303538
2,0.552795
3,0.578431
4,0.724138
5,0.200000
6,0.136364
7,0.333333
8,0.000000
11,0.000000


2) is_alone
เพราะคนที่มาคนเดียวอาจมี pattern การรอดต่างจากคนที่มากับครอบครัว boolean/0-1 แบบนี้โมเดลใช้ได้ง่าย

In [123]:
df["is_alone"] = (df["family_size"] == 1).astype(int)

In [124]:
df["is_alone"].value_counts()
df.groupby("is_alone")["survived"].mean()

,survived
is_alone,
0,0.505650
1,0.303538




**STEP 2: เลือก feature ชุดแรกสำหรับ baseline**

In [125]:
features = [
    "pclass",
    "sex",
    "age",
    "fare",
    "embarked",
    "family_size",
    "is_alone",
]

In [126]:
target = "survived"

**STEP 3: สร้างตารางใหม่สำหรับทำโมเดล**

In [127]:
model_df = df[features + [target]].copy()
model_df.head()

,pclass,sex,age,fare,embarked,family_size,is_alone,survived
0,3,male,22.0,7.2500,S,2,0,0
1,1,female,38.0,71.2833,C,2,0,1
2,3,female,26.0,7.9250,S,1,1,1
3,1,female,35.0,53.1000,S,2,0,1
4,3,male,35.0,8.0500,S,1,1,0


In [128]:
model_df.isnull().sum()

,0
pclass,0
sex,0
age,0
fare,0
embarked,0
family_size,0
is_alone,0
survived,0


**STEP 4: แปลง categorical ให้โมเดลใช้ได้**

ตอนนี้ sex กับ embarked ยังเป็นข้อความ
เราต้องแปลงก่อน

In [129]:
model_df = pd.get_dummies(model_df, columns=["sex", "embarked"], drop_first=True)
model_df.head()

,pclass,age,fare,family_size,is_alone,survived,sex_male,embarked_Q,embarked_S
0,3,22.0,7.2500,2,0,0,True,False,True
1,1,38.0,71.2833,2,0,1,False,False,False
2,3,26.0,7.9250,1,1,1,False,False,True
3,1,35.0,53.1000,2,0,1,False,False,True
4,3,35.0,8.0500,1,1,0,True,False,True


เหตุผล:

โมเดลส่วนใหญ่กินตัวเลข

get_dummies คือวิธีมาตรฐานสำหรับ categorical แบบง่าย

drop_first=True ลดความซ้ำซ้อนบางส่วน

**STEP 5: แยก X และ y**

In [130]:
X = model_df.drop("survived", axis=1)
y = model_df["survived"]

In [131]:
X.shape, y.shape

((891, 8), (891,))

STEP 6: เตรียม baseline model ตัวแรก

In [133]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, classification_report

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)

model = LogisticRegression(max_iter=1000)
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

print("Accuracy:", accuracy_score(y_test, y_pred))
print(classification_report(y_test, y_pred))

Accuracy: 0.7932960893854749
              precision    recall  f1-score   support

           0       0.82      0.85      0.84       110
           1       0.75      0.70      0.72        69

    accuracy                           0.79       179
   macro avg       0.78      0.78      0.78       179
weighted avg       0.79      0.79      0.79       179



Baseline model:
- ใช้ Logistic Regression เป็น baseline
- ใช้ feature หลัก เช่น pclass, sex, age, fare
- เพิ่ม engineered features คือ family_size และ is_alone